common/tor.c

```diff
@@ -0,0 +1,232 @@
+	tal_resize(&reach->ld->wireaddrs, n+1);
+	reach->ld->wireaddrs[n].type = ADDR_TYPE_TOR_V2;
+	reach->ld->wireaddrs[n].addrlen = TOR_V2_ADDRLEN;
+	reach->ld->wireaddrs[n].port = reach->ld->portnum;
+	b32_decode((u8 *)tor_dec_bytes,(u8 *)reach->buffer,2);
+	memcpy(&reach->ld->wireaddrs[n].addr,tor_dec_bytes, reach->ld->wireaddrs[n].addrlen);
+
+	}
+	/*on the other hand we can stay connected until ln finish to keep onion alive and then vanish*/
+	//because when we run with Detach flag as we now do every start of LN creates a new addr while the old
+	//stays valid until reboot this might not be desired so we can also drop Detach and use the
+	//read_partial to keep it open until LN drops
+	//FIXME: SAIBATO
+	//return io_read_partial(conn, reach->p, 1 ,&reach->hlen, io_tor_connect_create_onion_finished, reach);
+	return io_close(conn);
+}
+
+
+static struct io_plan *io_tor_connect_after_create_onion(struct io_conn *conn, struct tor_service_reaching *reach)
+{
+
+	reach->p = reach->p+reach->hlen;
+
+ if (!strstr((char *)reach->buffer, "ServiceID=")) {
+	if (reach->hlen == 0)
+		return io_close(conn);
+	return io_read_partial(conn, reach->p, 1 ,&reach->hlen, io_tor_connect_after_create_onion, reach);
+	}
+	else
+	{
+	memset(reach->buffer,0,sizeof(reach->buffer));
+	return io_read_partial(conn, reach->buffer, MAX_TOR_ONION_V2_ADDR_LEN ,&reach->hlen, io_tor_connect_create_onion_finished, reach);
+	};
+
+}
+
+
+//V3 tor after 3.3.3.aplha FIXME: TODO SAIBATO
+//sprintf((char *)reach->buffer,"ADD_ONION NEW:ED25519-V3 Port=9735,127.0.0.1:9735\r\n");
+
+static struct io_plan *io_tor_connect_make_onion(struct io_conn *conn, struct tor_service_reaching *reach)
+{
+	if (strstr((char *)reach->buffer, "250 OK") == NULL)
+				return io_close(conn);
+	sprintf((char *)reach->buffer, "ADD_ONION NEW:RSA1024 Port=%d,127.0.0.1:%d Flags=DiscardPK,Detach\r\n",
+			reach->ld->portnum,reach->ld->portnum);
+
+			reach->hlen = strlen((char *)reach->buffer);
+			reach->p = reach->buffer;
+	return io_write(conn, reach->buffer,  reach->hlen, io_tor_connect_after_create_onion, reach);
+
+}
+
+
+static struct io_plan *io_tor_connect_after_authenticate(struct io_conn *conn, struct tor_service_reaching *reach)
+{
+return io_read(conn, reach->buffer,7, io_tor_connect_make_onion, reach);
+}
+
+
+static struct io_plan *io_tor_connect_authenticate(struct io_conn *conn, struct tor_service_reaching *reach)
+{
+		sprintf((char *)reach->buffer, "AUTHENTICATE %s\r\n",(char *)reach->cookie);
+
+		if (reach->noauth)
+				sprintf((char *)reach->buffer, "AUTHENTICATE\r\n");
+
+		reach->hlen = strlen((char *)reach->buffer);
+
+return io_write(conn, reach->buffer, reach->hlen, io_tor_connect_after_authenticate, reach);
+
+}
+
+
+
+static struct io_plan *io_tor_connect_after_answer_pi(struct io_conn *conn, struct tor_service_reaching *reach)
+{
+	char *p,*p2;
+	int i;
+	static u8 buf[MAX_TOR_COOKIE_LEN];
+
+	reach->noauth=false;
+
+	if (strstr((char *)reach->buffer, "NULL")) reach->noauth=true;
```

Comment:
Is the buffer null-terminated under all conditions?  Does the server return null-terminated string or do we ensure the buffer is null-terminated somehow...?  Also should be:

    if (whatever)
            reach->noauth = true;